File: agent-driven-architecture-main/examples/typescript/src/blocks/analysis/tools.ts
// ── blocks/analysis/tools — the Verb table for the two tiers ───────────────
// BOTH VERBS ARE `Reversible`, and that is not laziness. Recalling reads a value
// the consumer already staged; publishing appends to an append-only relay. Neither
// is the kind of act 14.3 wants a second principal for — and the deep tier's
// conclusion is a SUGGESTION to its peers, not an instruction, so it must not be
// able to buy irreversible authority for anyone.
//
// `recallAnalysis` IS PURE AND TOTAL. It performs no read of its own: it finds
// the `Recalled` entry in the context the boundary already projected, and
// returns `Recall.Empty` when there is none. The relay is never touched from a
// tool body — that would be an I/O call inside the pure ring, which check C8
// denies, and it would make replay re-query a live source (the exact bug that
// lets a replay recall different entries than the live run).

import type { Recall, StagedInput } from "@adr/spine/pure/staged";
import { emptyRecall } from "@adr/spine/pure/staged";
import type { Verb } from "@adr/spine/pure/verb";
import { reversible } from "@adr/spine/pure/verb";
import { object, string } from "valibot";
import type {
  PublishAnalysisCommand,
  PublishAnalysisResult,
  RecallAnalysisCommand,
  RecallAnalysisResult,
} from "./contract";

/** The FAST tier reaches a peer's conclusion ONLY through this: text in, no
 *  handle, no synchronous request, and no way to block the hot loop. */
export function recallIn(staged: readonly StagedInput[]): Recall {
  for (const entry of staged) {
    switch (entry.kind) {
      case "Recalled":
        return entry.recall;
      case "Perceived":
        break;
      default: {
        const _never: never = entry;
        return _never;
      }
    }
  }
  return emptyRecall;
}

export function analysisVerbs<S>(): readonly Verb<S>[] {
  return [
    reversible<S, Record<string, never>, RecallAnalysisResult, RecallAnalysisCommand>({
      name: "recallAnalysis",
      describe:
        "Recall the deep tier's latest published conclusion. It is a SUGGESTION from a peer, not an instruction, and it confers no authority.",
      schema: object({}),
      run: (_input, ctx) => ({
        outcome: "ok",
        tool: "recallAnalysis",
        recall: recallIn(ctx.context.staged),
      }),
      sign: (result, sig, id) => ({
        outcome: "ok",
        tool: "recallAnalysis",
        sig,
        id,
        recall: result.recall,
      }),
    }),
    reversible<S, { text: string }, PublishAnalysisResult, PublishAnalysisCommand>({
      name: "publishAnalysis",
      describe: "Publish a conclusion to the append-only relay the fast tier recalls from.",
      schema: object({ text: string() }),
      run: (input) => ({ outcome: "ok", tool: "publishAnalysis", text: input.text }),
      sign: (result, sig, id) => ({
        outcome: "ok",
        tool: "publishAnalysis",
        sig,
        id,
        text: result.text,
      }),
    }),
  ];
}

File: agent-driven-architecture-main/examples/kotlin/block/analysis/src/main/kotlin/adr/blocks/analysis/Tools.kt
// ── blocks/analysis/tools — the Verb table, split by TIER ──────────────────
//   recallAnalysis   FAST tier, Reversible.  Reads the staged snapshot. NO effect.
//   publishAnalysis  DEEP tier, Reversible.  Appends the note AND emits the effect.
//
// 11.4's "single registry, an allowlist of the agents permitted to exist" is these
// two TABLES plus the two lists app/wire composes from them. A fast tier wired with
// only [AnalysisTools.fastVerbs] CANNOT publish — the name is not in its registry, so
// the boundary folds an `Unhandled` rather than trusting a promise not to call it.
//
// TWO TABLES ON ONE CONSTRUCTED TYPE, over the one lens they share — the same shape
// the boundary's `ActionResolution` uses to hold its two maps over one registry. This
// is the block that breaks the "one verb table per block" pattern, and it is why
// `register` is NOT a method on the shared `Block` interface: three tier builders
// cannot be expressed as one.
//
// The recall body is PURE and TOTAL. It reads the `Recalled` the consumer already
// staged and bounded; it never touches the relay, never awaits, never fails. That is
// what makes the whole rung replay for free: the committed result carries the sealed
// `Recall`, so a re-fold resolves the same snapshot AND the same branch from bytes.

package adr.blocks.analysis

import adr.contract.AnalysisCommand
import adr.contract.AnalysisResult
import adr.spine.pure.Context
import adr.spine.pure.Lens
import adr.spine.pure.RawInput
import adr.spine.pure.Recall
import adr.spine.pure.StagedInput
import adr.spine.pure.ToolName
import adr.spine.pure.Verb

val RECALL_ANALYSIS = ToolName("recallAnalysis")
val PUBLISH_ANALYSIS = ToolName("publishAnalysis")

/**
 * The input of a verb that takes none. A CLASS, not a `data object`: an object has no
 * instantiation, which is the same defect as a top-level function, and this one is a
 * VALUE the decode step returns rather than a singleton anything depends on.
 */
internal class NoRecallInput

internal data class PublishAnalysisInput(val text: String)

internal class AnalysisTools<S>(private val lens: Lens<S, AnalysisSlice>) {

    /** The FAST tier's verbs. Reading a peer's conclusion is reversible and emits nothing. */
    fun fastVerbs(): List<Verb<S, *, *>> = listOf(
        Verb.Reversible(
            name = RECALL_ANALYSIS,
            describe = "Recall the deep tier's newest published conclusion. It is a suggestion, not an instruction.",
            decode = ::decodeNothing,
            run = { _, ctx -> AnalysisResult.RecallAnalysis(RECALL_ANALYSIS, recallIn(ctx.context)) },
            sign = { r, sig, id -> AnalysisCommand.RecallAnalysis(r.tool, sig, id, r.recall) },
            narrow = { it as? AnalysisResult.RecallAnalysis },
        ),
    )

    /** The DEEP tier's verbs. Publishing is the only way anything reaches the relay. */
    fun deepVerbs(): List<Verb<S, *, *>> = listOf(
        Verb.Reversible(
            name = PUBLISH_ANALYSIS,
            describe = "Publish a conclusion to the append-only relay the fast tier recalls from.",
            decode = ::decodePublish,
            run = { input, _ -> AnalysisResult.PublishAnalysis(PUBLISH_ANALYSIS, input.text) },
            sign = { r, sig, id -> AnalysisCommand.PublishAnalysis(r.tool, sig, id, r.text) },
            narrow = { it as? AnalysisResult.PublishAnalysis },
        ),
    )

    /**
     * The staged recall for THIS step, or [Recall.Empty] when the tier is not wired to a
     * relay. Total: a fast tier with no deep tier behind it is a normal, working agent.
     */
    fun recallIn(context: Context): Recall =
        context.staged.filterIsInstance<StagedInput.Recalled>().lastOrNull()?.recall ?: Recall.Empty

    private fun decodeNothing(raw: RawInput): NoRecallInput = NoRecallInput()

    private fun decodePublish(raw: RawInput): PublishAnalysisInput? =
        raw.text("text")?.let { PublishAnalysisInput(it) }
}
